File: main_project/tripod-new/src/cpp/configclient/ConfigClientFactory.cpp
#include <Common.h>
#include <CacheManagerConfigClient.h>
#include <ProducerManagerConfigClient.h>
#include <CacheConfigClient.h>
#include <ConfigClientFactory.h>
#include <CacheConfigClient.h>
#include <RedisCacheConfigClient.h>

namespace com {
namespace renren {
namespace tripod { 

CacheManagerConfigClientPtr ConfigClientFactory::getCacheManagerConfigClient(const std::string& zkAddress) {
    MCE_INFO("ConfigClientFactory::getCacheManagerConfigClient zkAddress" << zkAddress);
    std::map<std::string, CacheManagerConfigClientPtr>::iterator findIt;
    {
        boost::shared_lock<boost::shared_mutex> lock(cacheManagerConfigClientMaplock_);
        findIt = cacheManagerConfigClientMap_.find(zkAddress);
        if( findIt != cacheManagerConfigClientMap_.end() ){
            return findIt->second;
        }
    }
    {
        boost::upgrade_lock<boost::shared_mutex> lock(cacheManagerConfigClientMaplock_);
        boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock1(lock);
        findIt = cacheManagerConfigClientMap_.find(zkAddress);
        if( findIt != cacheManagerConfigClientMap_.end() ){
            return findIt->second;
        }
        CacheManagerConfigClientPtr cmc(new CacheManagerConfigClient(zkAddress));
        if ( cmc->initialize() ) {
            MCE_INFO("ConfigClientFactory::getCacheManagerConfigClient init for zkAddress" << zkAddress << " success!");
            cacheManagerConfigClientMap_.insert(make_pair(zkAddress, cmc));
            return cmc;
        }
        MCE_INFO("ConfigClientFactory::getCacheManagerConfigClient init for zkAddress" << zkAddress << " failed!");
        return CacheManagerConfigClientPtr();
    }
}

ProducerManagerConfigClientPtr ConfigClientFactory::getProducerManagerConfigClient(const std::string& zkAddress) {
    MCE_INFO("ConfigClientFactory::getProducerManagerConfigClient zkAddress" << zkAddress);
    std::map<std::string, ProducerManagerConfigClientPtr>::iterator findIt;
    {
        boost::shared_lock<boost::shared_mutex> lock(producerManagerConfigClientMaplock_);
        findIt = producerManagerConfigClientMap_.find(zkAddress);
        if( findIt != producerManagerConfigClientMap_.end() ){
            return findIt->second;
        }
    }
    {
        boost::upgrade_lock<boost::shared_mutex> lock(producerManagerConfigClientMaplock_);
        boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock1(lock);
        findIt = producerManagerConfigClientMap_.find(zkAddress);
        if( findIt != producerManagerConfigClientMap_.end() ){
            return findIt->second;
        }
        ProducerManagerConfigClientPtr pmc(new ProducerManagerConfigClient(zkAddress));
        if ( pmc->initialize() ) {
            MCE_INFO("ConfigClientFactory::getProducerManagerConfigClient init for zkAddress" << zkAddress << " success!");
            producerManagerConfigClientMap_.insert(make_pair(zkAddress, pmc));
            return pmc;
        }
        MCE_INFO("ConfigClientFactory::getProducerManagerConfigClient init for zkAddress" << zkAddress << " failed!");
        return ProducerManagerConfigClientPtr();
    }
}

CacheConfigClientPtr ConfigClientFactory::getCacheConfigClient(const std::string& zkAddress) {
    MCE_INFO("ConfigClientFactory::getCacheConfigClient zkAddress" << zkAddress);
    std::map<std::string, CacheConfigClientPtr>::iterator findIt;
    {
        boost::shared_lock<boost::shared_mutex> lock(cacheConfigClientMaplock_);
        findIt = cacheConfigClientMap_.find(zkAddress);
        if( findIt != cacheConfigClientMap_.end() ){
            return findIt->second;
        }
    }
    {
        boost::upgrade_lock<boost::shared_mutex> lock(cacheConfigClientMaplock_);
        boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock1(lock);
        findIt = cacheConfigClientMap_.find(zkAddress);
        if( findIt != cacheConfigClientMap_.end() ){
            return findIt->second;
        }
        CacheConfigClientPtr cc(new CacheConfigClient(zkAddress));
        if ( cc->initialize() ) {
            MCE_INFO("ConfigClientFactory::getCacheConfigClient init for zkAddress" << zkAddress << " success!");
            cacheConfigClientMap_.insert(make_pair(zkAddress, cc));
            return cc;
        }
        MCE_INFO("ConfigClientFactory::getCacheConfigClient init for zkAddress" << zkAddress << " failed!");
        return CacheConfigClientPtr();
    }
}

RedisCacheConfigClientPtr ConfigClientFactory::GetRedisCacheConfigClient(const std::string& zk_address) {
  MCE_INFO("ConfigClientFactory::GetRedisCacheConfigClient zk_address" << zk_address);
  MCE_DEBUG("===============ConfigClientFactory::GetRedisCacheConfigClient zk_address" << zk_address);
  std::map<std::string, RedisCacheConfigClientPtr>::iterator findIt;
  {
      boost::shared_lock<boost::shared_mutex> lock(redis_cache_config_client_map_lock_);
      findIt = redis_cache_config_client_map_.find(zk_address);
      if( findIt != redis_cache_config_client_map_.end() ){
          return findIt->second;
      }
  }
  {
      boost::upgrade_lock<boost::shared_mutex> lock(redis_cache_config_client_map_lock_);
      boost::upgrade_to_unique_lock<boost::shared_mutex> uniqueLock1(lock);
      findIt = redis_cache_config_client_map_.find(zk_address);
      if( findIt != redis_cache_config_client_map_.end() ){
          return findIt->second;
      }
      RedisCacheConfigClientPtr cc(new RedisCacheConfigClient(zk_address));
      if ( cc->initialize() ) {
          MCE_INFO("liyong ConfigClientFactory::GetRedisCacheConfigClient init for zk_address" << zk_address << " success!");
          redis_cache_config_client_map_.insert(make_pair(zk_address, cc));
          return cc;
      }
      MCE_INFO("liyong ConfigClientFactory::GetRedisCacheConfigClient init for zk_address" << zk_address << " failed!");
      return RedisCacheConfigClientPtr();
  }
}
    
}
}
}

File: main_project/tripod-new-branch/src/java/TripodConfig/src/main/java/com/renren/tripod/configclient/ConfigClientFactory.java
package com.renren.tripod.configclient;

import java.util.concurrent.ConcurrentHashMap;

import com.renren.tripod.configutil.ConfigData;

public class ConfigClientFactory {
    private static org.apache.commons.logging.Log _logger = org.apache.commons.logging.LogFactory
	    .getLog(ConfigClientFactory.class);
    private static ConcurrentHashMap<String, CacheManagerConfigClient> _cacheManagerConfigClientMap = new ConcurrentHashMap<String, CacheManagerConfigClient>();
    private static ConcurrentHashMap<String, ProducerManagerConfigClient> _producerManagerConfigClientMap = new ConcurrentHashMap<String, ProducerManagerConfigClient>();
    private static ConcurrentHashMap<String, CacheConfigClient> _cacheConfigClientMap = new ConcurrentHashMap<String, CacheConfigClient>();

    //
    private static ConcurrentHashMap<String, RedisCacheConfigClient> redisCacheConfigClientMap = new ConcurrentHashMap<String, RedisCacheConfigClient>();

    public static synchronized CacheManagerConfigClient getCacheMangerConfigClient(
	    String zkAddress) {
	_logger.info("ConfigClientFactory.getCacheManagerConfigClient zkAddress "
		+ zkAddress);
	CacheManagerConfigClient cc = _cacheManagerConfigClientMap
		.get(zkAddress);
	if (null == cc) {
	    _logger.info("ConfigClientFactory.getCacheManagerConfigClient new CacheManagerConfigClient for zkAddress "
		    + zkAddress);
	    cc = new CacheManagerConfigClient(zkAddress,
		    ConfigData.TRIPOD_MANAGER);
	    if (cc.initialize()) {
		_cacheManagerConfigClientMap.put(zkAddress, cc);
		return cc;
	    } else {
		return null;
	    }
	}
	return cc;
    }

    public static synchronized ProducerManagerConfigClient getProducerMangerConfigClient(
	    String zkAddress) {
	_logger.info("ConfigClientFactory.getProducerManagerConfigClient zkAddress "
		+ zkAddress);
	ProducerManagerConfigClient cc = _producerManagerConfigClientMap
		.get(zkAddress);
	if (null == cc) {
	    _logger.info("ConfigClientFactory.getProducerManagerConfigClient new ProducerManagerConfigClient for zkAddress "
		    + zkAddress);
	    cc = new ProducerManagerConfigClient(zkAddress,
		    ConfigData.TRIPOD_PRODUCER);
	    if (cc.initialize()) {
		_producerManagerConfigClientMap.put(zkAddress, cc);
		return cc;
	    } else {
		return null;
	    }
	}
	return cc;
    }

    public static synchronized CacheConfigClient getCacheConfigClient(
	    String zkAddress) {
	_logger.info("ConfigClientFactory.getCacheConfigClient zkAddress "
		+ zkAddress);
	CacheConfigClient cc = _cacheConfigClientMap.get(zkAddress);
	if (null == cc) {
	    _logger.info("ConfigClientFactory.getCacheConfigClient new CacheConfigClient for zkAddress "
		    + zkAddress);
	    cc = new CacheConfigClient(zkAddress, ConfigData.TRIPOD_CACHE);

	    if (cc.initialize()) {
		_cacheConfigClientMap.put(zkAddress, cc);
		return cc;
	    } else {
		return null;
	    }
	}
	return cc;
    }

    public static synchronized RedisCacheConfigClient getRedisCacheConfigClient(
	    String zkAddress) {
	_logger.info("ConfigClientFactory.getRedisCacheConfigClient zkAddress "
		+ zkAddress);
	RedisCacheConfigClient configClient = redisCacheConfigClientMap
		.get(zkAddress);
	if (configClient != null) {
	    return configClient;
	} else {
	    _logger.info("ConfigClientFactory.getRedisCacheConfigClient new RedisCacheConfigClient for zkAddress "
		    + zkAddress);
	    configClient = new RedisCacheConfigClient(zkAddress,
		    ConfigData.TRIPOD_CACHE);
	    if (configClient.initialize()) {
		redisCacheConfigClientMap.put(zkAddress, configClient);
		return configClient;
	    } else {
		return null;
	    }
	}
    }

}